File: client-sdks/elixir/tests/comm/socket_helper_test.exs
# ------------------------------------------------------------------------------
# Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------------

defmodule VertexCacheSdk.Comm.SocketHelperTest do
  use ExUnit.Case
  alias VertexCacheSdk.Comm.SocketHelper
  alias VertexCacheSdk.Model.ClientOption
  alias VertexCacheSdk.Model.VertexCacheSdkException

  @unused_port 59999
  @live_tls_port 50505
  @mock_port 18888

  setup_all do
    {:ok, listener} =
      :gen_tcp.listen(@mock_port, [:binary, packet: :raw, active: false, reuseaddr: true])

    spawn(fn -> accept_loop(listener) end)
    on_exit(fn -> :gen_tcp.close(listener) end)
    :ok
  end

  defp accept_loop(listener) do
    case :gen_tcp.accept(listener) do
      {:ok, socket} ->
        :gen_tcp.close(socket)
        accept_loop(listener)

      _ ->
        :ok
    end
  end

  test "create_socket_non_tls should succeed" do
    option = %ClientOption{
      server_host: "localhost",
      server_port: @mock_port,
      connect_timeout: 1000,
      read_timeout: 1000
    }

    case SocketHelper.create_socket_non_tls(option) do
      {:ok, socket} ->
        assert is_port(socket)
        :gen_tcp.close(socket)

      {:error, err} ->
        flunk("Expected success, got error: #{err.message}")
    end
  end

  test "create_socket_non_tls should fail if port closed" do
    option = %ClientOption{
      server_host: "localhost",
      server_port: @unused_port,
      connect_timeout: 500,
      read_timeout: 500
    }

    case SocketHelper.create_socket_non_tls(option) do
      {:error, %VertexCacheSdkException{message: msg}} ->
        assert msg == "Failed to create Non Secure Socket"

      other ->
        flunk("Expected error, got: #{inspect(other)}")
    end
  end

  test "create_socket_non_tls should fail on timeout" do
    option = %ClientOption{
      server_host: "10.255.255.1",
      server_port: 9999,
      connect_timeout: 1,
      read_timeout: 100
    }

    case SocketHelper.create_socket_non_tls(option) do
      {:error, %VertexCacheSdkException{message: msg}} ->
        assert msg == "Failed to create Non Secure Socket"

      other ->
        flunk("Expected timeout error, got: #{inspect(other)}")
    end
  end

  test "create_secure_socket should fail due to missing TLS context" do
    option = %ClientOption{
      server_host: "localhost",
      server_port: 8443,
      connect_timeout: 1000,
      read_timeout: 1000,
      verify_certificate: true
    }

    case SocketHelper.create_secure_socket(option) do
      {:error, %VertexCacheSdkException{message: msg}} ->
        assert msg == "Failed to create Secure Socket"

      other ->
        flunk("Expected TLS context error, got: #{inspect(other)}")
    end
  end

  test "create_secure_socket should fail with bad certificate" do
    option = %ClientOption{
      server_host: "localhost",
      server_port: @mock_port,
      connect_timeout: 1000,
      read_timeout: 1000,
      verify_certificate: true,
      tls_certificate: "not a valid cert"
    }

    case SocketHelper.create_secure_socket(option) do
      {:error, %VertexCacheSdkException{message: msg}} ->
        assert msg == "Failed to create Secure Socket"

      other ->
        flunk("Expected cert error, got: #{inspect(other)}")
    end
  end

  @tag :tls
  test "create_secure_socket should succeed with live TLS server (verify_certificate: false)" do
    unless System.get_env("ENABLE_LIVE_TLS_TESTS") == "true" do
      IO.puts("Live TLS test skipped: ENABLE_LIVE_TLS_TESTS not set to 'true'")
      assert true
    else
      option = %ClientOption{
        server_host: "localhost",
        server_port: @live_tls_port,
        connect_timeout: 1000,
        read_timeout: 1000,
        verify_certificate: false
      }

      case SocketHelper.create_secure_socket(option) do
        {:ok, ssl_socket} ->
          assert is_pid(ssl_socket)
          :ssl.close(ssl_socket)

        {:error, err} ->
          IO.warn("Live TLS server not reachable or handshake failed: #{err.message}")
          assert true
      end
    end
  end

end

File: client-sdks/go/tests/comm/socket_helper_test.go
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------------

package comm_test

import (
	"github.com/vertexcache/vertexcache/client-sdks/go/sdk/comm"
	"github.com/vertexcache/vertexcache/client-sdks/go/sdk/model"
	"net"
	"os"
	"testing"
)

const (
	validPort   = 11211
	unusedPort  = 59999
	liveTLSPort = 50505
	tcpTimeout  = 1000
	readTimeout = 1000
)

func TestCreateSocketNonTLSShouldSucceed(t *testing.T) {
	// Start a temporary TCP server
	ln, err := net.Listen("tcp", "localhost:0") // OS assigns an available port
	if err != nil {
		t.Fatalf("Failed to start temporary TCP server: %s", err)
	}
	defer ln.Close()

	// Get the assigned port
	addr := ln.Addr().(*net.TCPAddr)
	validPort := addr.Port

	// Run server handler in background (accept and close)
	go func() {
		conn, err := ln.Accept()
		if err == nil {
			conn.Close()
		}
	}()

	opt := model.ClientOption{
		ServerHost:     "localhost",
		ServerPort:     validPort,
		ConnectTimeout: tcpTimeout,
		ReadTimeout:    readTimeout,
	}

	sock, err := comm.CreateSocketNonTLS(opt)
	if err != nil {
		t.Fatalf("Expected success, got error: %s", err.Error())
	}
	sock.Close()
}

func TestCreateSocketNonTLSShouldFailIfPortClosed(t *testing.T) {
	opt := model.ClientOption{
		ServerHost:     "localhost",
		ServerPort:     unusedPort,
		ConnectTimeout: tcpTimeout,
		ReadTimeout:    readTimeout,
	}
	_, err := comm.CreateSocketNonTLS(opt)
	if err == nil {
		t.Fatal("Expected error, got success")
	}
}

func TestCreateSocketNonTLSShouldFailOnTimeout(t *testing.T) {
	opt := model.ClientOption{
		ServerHost:     "10.255.255.1",
		ServerPort:     9999,
		ConnectTimeout: 1,
		ReadTimeout:    readTimeout,
	}
	_, err := comm.CreateSocketNonTLS(opt)
	if err == nil {
		t.Fatal("Expected timeout error, got success")
	}
}

func TestCreateSecureSocketShouldFailDueToMissingTLSContext(t *testing.T) {
	opt := model.ClientOption{
		ServerHost:        "localhost",
		ServerPort:        8443,
		ConnectTimeout:    tcpTimeout,
		ReadTimeout:       readTimeout,
		VerifyCertificate: true,
	}
	_, err := comm.CreateSecureSocket(opt)
	if err == nil {
		t.Fatal("Expected TLS context error, got success")
	}
}

func TestCreateSecureSocketShouldFailWithBadCertificate(t *testing.T) {
	opt := model.ClientOption{
		ServerHost:        "localhost",
		ServerPort:        18888,
		ConnectTimeout:    tcpTimeout,
		ReadTimeout:       readTimeout,
		VerifyCertificate: true,
		TLSCertificate:    "not a valid cert",
	}
	_, err := comm.CreateSecureSocket(opt)
	if err == nil {
		t.Fatal("Expected cert error, got success")
	}
}

func TestCreateSecureSocketShouldSucceedWithLiveServer(t *testing.T) {
	if os.Getenv("VC_LIVE_TEST") != "true" {
		t.Skip("Skipping live test unless VC_LIVE_TEST=true")
	}

	opt := model.ClientOption{
		ServerHost:        "localhost",
		ServerPort:        liveTLSPort,
		ConnectTimeout:    tcpTimeout,
		ReadTimeout:       readTimeout,
		VerifyCertificate: false,
	}

	sock, err := comm.CreateSecureSocket(opt)
	if err != nil {
		t.Logf("Live TLS test skipped: could not establish secure socket: %s", err.Error())
		return // Don't fail test — treat as a soft skip
	}

	defer sock.Close()

	t.Log("TLS connection established successfully (expected IDENT failure may follow).")
}
